**Janani Jayashankar**  - **ENG19CS0125**

**Kehkashaan Khanum A** - **ENG19CS0142**

**Kumuda Kanatti**      - **ENG19CS0149**

In [1]:
import pandas as pd
import numpy as np

In [2]:
data = pd.read_csv('Restaurant_Reviews.tsv', sep = '\t')
data.head(10)

Review  Liked
0                           Wow... Loved this place.      1
1                                 Crust is not good.      0
2          Not tasty and the texture was just nasty.      0
3  Stopped by during the late May bank holiday of...      1
4  The selection on the menu was great and so wer...      1
5     Now I am getting angry and I want my damn pho.      0
6              Honeslty it didn't taste THAT fresh.)      0
7  The potatoes were like rubber and you could te...      0
8                          The fries were great too.      1
9                                     A great touch.      1

In [3]:
data['Liked'].value_counts()

1    500
0    500
Name: Liked, dtype: int64

In [4]:
data.shape

(1000, 2)

In [5]:
data.describe()

Liked
count  1000.00000
mean      0.50000
std       0.50025
min       0.00000
25%       0.00000
50%       0.50000
75%       1.00000
max       1.00000

## Data Processing

In [6]:
import nltk
import re
from nltk.corpus import stopwords 
from nltk.stem.porter import PorterStemmer
ps = PorterStemmer()

In [7]:
stop = stopwords.words('english')

In [8]:
cleaned_review = []

for i in range(len(data)):
    review = re.sub('[^a-zA-Z]', ' ', data['Review'][i])    
    review = review.lower()    
    review = review.split()    
    review = [ps.stem(word) for word in review if word not in stop]    
    review = " ".join(review)    
    cleaned_review.append(review)
#print(cleaned_review)

In [9]:
from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer(max_features = 1500, ngram_range = (1,2))
X = cv.fit_transform(cleaned_review).toarray()
y = data.iloc[:,1].values # df['Liked']

In [10]:
cv.get_feature_names() # Use of n_grams in the range of 1, 2 

C:\Users\Kehkashaan Khanum A\anaconda3\lib\site-packages\sklearn\utils\deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


['absolut',
 'absolut amaz',
 'acknowledg',
 'actual',
 'ad',
 'ago',
 'almost',
 'also',
 'also serv',
 'also tast',
 'although',
 'alway',
 'alway great',
 'alway hit',
 'amaz',
 'ambianc',
 'ambienc',
 'amount',
 'anoth',
 'anoth minut',
 'anoth said',
 'anyon',
 'anyth',
 'anytim',
 'anytim soon',
 'anyway',
 'apolog',
 'appet',
 'area',
 'around',
 'around like',
 'arriv',
 'ask',
 'assur',
 'ate',
 'atmospher',
 'atmospher fun',
 'attack',
 'attent',
 'attent provid',
 'attitud',
 'authent',
 'authent thai',
 'averag',
 'averag best',
 'avoid',
 'avoid place',
 'aw',
 'away',
 'awesom',
 'awesom servic',
 'babi',
 'bachi',
 'bachi burger',
 'back',
 'back anytim',
 'back everi',
 'bacon',
 'bad',
 'bad food',
 'bagel',
 'bakeri',
 'bar',
 'bar food',
 'bare',
 'bartend',
 'basic',
 'bathroom',
 'batter',
 'bay',
 'bean',
 'beat',
 'beauti',
 'becom',
 'beef',
 'beef sandwich',
 'beer',
 'beer select',
 'behind',
 'believ',
 'belli',
 'best',
 'best breakfast',
 'best place',
 'be

In [11]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state = 0)

In [12]:
X_train.shape, X_test.shape

((800, 1500), (200, 1500))

# Building Models and Predicting Accuracy

In [13]:
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
from sklearn.naive_bayes import GaussianNB
clf1 = GaussianNB()
clf1.fit(X_train, y_train)
y_pred1 = clf1.predict(X_test)
print('Accuracy for Gaussian Naive Bayes: ', accuracy_score(y_test, y_pred1)*100)
print()
print(confusion_matrix(y_test, y_pred1))
print()
print("Classification Report for Gaussian Naive Bayes: ", classification_report(y_test, y_pred1))

Accuracy for Gaussian Naive Bayes:  73.0

[[58 39]
 [15 88]]

Classification Report for Gaussian Naive Bayes:                precision    recall  f1-score   support

           0       0.79      0.60      0.68        97
           1       0.69      0.85      0.77       103

    accuracy                           0.73       200
   macro avg       0.74      0.73      0.72       200
weighted avg       0.74      0.73      0.73       200



In [14]:
from sklearn.neighbors import KNeighborsClassifier
clf2 = KNeighborsClassifier()
clf2.fit(X_train, y_train)
y_pred2 = clf2.predict(X_test)
accuracy_score(y_test, y_pred2)
print('Accuracy for KNN: ', accuracy_score(y_test, y_pred2)*100)
print()
print(confusion_matrix(y_test, y_pred2))
print()
print("Classification Report for KNN: ", classification_report(y_test, y_pred2))

Accuracy for KNN:  62.5

[[91  6]
 [69 34]]

Classification Report for KNN:                precision    recall  f1-score   support

           0       0.57      0.94      0.71        97
           1       0.85      0.33      0.48       103

    accuracy                           0.62       200
   macro avg       0.71      0.63      0.59       200
weighted avg       0.71      0.62      0.59       200



In [15]:
from sklearn.linear_model import LogisticRegression
clf3 = LogisticRegression()
clf3.fit(X_train, y_train)
y_pred3 = clf3.predict(X_test)
print('Accuracy for Logistic Regression: ', accuracy_score(y_test, y_pred3)*100)
print()
print(confusion_matrix(y_test, y_pred3))
print()
print("Classification Report for Logistic Regression: ", classification_report(y_test, y_pred3))

Accuracy for Logistic Regression:  73.5

[[78 19]
 [34 69]]

Classification Report for Logistic Regression:                precision    recall  f1-score   support

           0       0.70      0.80      0.75        97
           1       0.78      0.67      0.72       103

    accuracy                           0.73       200
   macro avg       0.74      0.74      0.73       200
weighted avg       0.74      0.73      0.73       200



In [16]:
from sklearn.svm import SVC
clf4 = SVC()
clf4.fit(X_train, y_train)
y_pred4 = clf4.predict(X_test)
print('Accuracy for Support Vector Classifier: ', accuracy_score(y_test, y_pred4)*100)
print()
print(confusion_matrix(y_test, y_pred4))
print()
print("Classification Report for Support Vector Classifier: ", classification_report(y_test, y_pred4))

Accuracy for Support Vector Classifier:  75.5

[[88  9]
 [40 63]]

Classification Report for Support Vector Classifier:                precision    recall  f1-score   support

           0       0.69      0.91      0.78        97
           1       0.88      0.61      0.72       103

    accuracy                           0.76       200
   macro avg       0.78      0.76      0.75       200
weighted avg       0.78      0.76      0.75       200



In [17]:
from sklearn.tree import DecisionTreeClassifier
clf5 = DecisionTreeClassifier()
clf5.fit(X_train, y_train)
y_pred5 = clf5.predict(X_test)
print('Accuracy for Decision Tree Classifier: ', accuracy_score(y_test, y_pred5)*100)
print()
print(confusion_matrix(y_test, y_pred5))
print()
print("Classification Report for Decision Tree Classifier: ", classification_report(y_test, y_pred5))

Accuracy for Decision Tree Classifier:  66.0

[[64 33]
 [35 68]]

Classification Report for Decision Tree Classifier:                precision    recall  f1-score   support

           0       0.65      0.66      0.65        97
           1       0.67      0.66      0.67       103

    accuracy                           0.66       200
   macro avg       0.66      0.66      0.66       200
weighted avg       0.66      0.66      0.66       200



In [18]:
from sklearn.ensemble import RandomForestClassifier
clf6 = RandomForestClassifier()
clf6.fit(X_train, y_train)
y_pred6 = clf6.predict(X_test)
print('Accuracy for Random Forest Classifier: ', accuracy_score(y_test, y_pred6)*100)
print()
print(confusion_matrix(y_test, y_pred6))
print()
print("Classification Report for Random Forest Classifier: ", classification_report(y_test, y_pred6))

Accuracy for Random Forest Classifier:  70.5

[[77 20]
 [39 64]]

Classification Report for Random Forest Classifier:                precision    recall  f1-score   support

           0       0.66      0.79      0.72        97
           1       0.76      0.62      0.68       103

    accuracy                           0.70       200
   macro avg       0.71      0.71      0.70       200
weighted avg       0.71      0.70      0.70       200



By the use of Different Classifiers, we can say that the **Support Vector Classifier**  is the best for Restaurant Review Analysis with **75.50%** Accuracy.